### PR TITLE
Add Eventhubs Perf Test.

### PR DESCRIPTION
**Pull request by g2vinay**
Adds Perf test for Track 2 and Track 1 EventHubs.
For Send and Receive events scenario.

### CODE REVIEW

File: sdk/storage/azure-storage-perf/src/main/java/com/azure/storage/App.java [hunk 1]
```diff
@@ -1,9 +1,10 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-package com.azure.storage.blob.perf;
+package com.azure.storage;
```
Comment (g2vinay) [inline sdk/storage/azure-storage-perf/src/main/java/com/azure/storage/App.java:4]: revert this change.

File: sdk/eventhubs/azure-messaging-eventhubs-perf/pom.xml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  <groupId>com.azure</groupId>
+  <artifactId>azure-messaging-eventhubs-perf</artifactId>
+  <version>1.0.0-beta.1</version>
+
+
+  <dependencies>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-messaging-eventhubs</artifactId>
+      <version>5.3.1</version> <!-- {x-version-update;com.azure:azure-messaging-eventhubs;current} -->
```
Comment (g2vinay) [inline sdk/eventhubs/azure-messaging-eventhubs-perf/pom.xml:23]: map it to source version.

File: sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/standalone/EventHubsConsumerClientPerf.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     *
+     * @param args The arguments used to run the performance test.
+     * @throws InterruptedException when thread is interrupted during execution.
+     */
+    public static void main(String[] args) throws InterruptedException {
+        Options options = new Options();
+        Option clientsOption = new Option("c", "clients", true, "Number of client instances");
+        options.addOption(clientsOption);
+        Option partitionsOption = new Option("p", "partitions", true, "Number of partitions");
+        options.addOption(partitionsOption);
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/standalone/EventHubsConsumerClientPerf.java:41]: Not sure why this option is required. The number of partitions for an event hub is defined on the service side.

File: sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/SendEventBatchPartitionTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        if (options.getPartitionKey() != null) {
+            batchOptions.partitionKey = options.getPartitionKey();
+        }
+
+        eventDataBatch = partitionSender.createBatch(batchOptions);
+
+        for (int i = 0; i < options.getEvents(); i++) {
+            if (!eventDataBatch.tryAdd(EventData.create("Static Event".getBytes(StandardCharsets.UTF_8)))) {
```
Comment (srnagar) [inline sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/SendEventBatchPartitionTest.java:45]: same here - creation of event data can be done just once.

File: sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/SendEventBatchPartitionTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import com.microsoft.azure.eventhubs.perf.core.EventHubsPerfStressOptions;
+import com.microsoft.azure.eventhubs.perf.core.ServiceTest;
+import reactor.core.publisher.Mono;
+
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Runs the Send Events Batch Performance Test for EventHubs.
+ */
+public class SendEventBatchPartitionTest extends ServiceTest<EventHubsPerfStressOptions> {
```
Comment (srnagar) [inline sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/SendEventBatchPartitionTest.java:19]: Do we need separate classes for testing "send to specific partition" vs "send to any partition"?

File: sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/SendEventBatchTest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     * Runs the Send Events in Batch peformance test.
+     *
+     * @param options The options bag to use for testing.
+     * @throws IllegalStateException when the specified number of messages cannot fit in the specified batch size.
+     */
+    public SendEventBatchTest(EventHubsPerfStressOptions options) throws IllegalStateException {
+        super(options);
+        CreateBatchOptions createBatchOptions = getBatchOptions(options);
+        if (createBatchOptions != null) {
+            eventDataBatch = eventHubProducerClient.createBatch(createBatchOptions);
```
[the remark this answers is not in the document]
Comment (mikeharder) [reply sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/SendEventBatchTest.java:29]: If this code is making a network request, the `setupAsync()` method would be a better place than the constructor.  The setup and cleanup methods are async-only, and it shouldn't impact throughput if you use the async client in setup/cleanup and the sync client in `run()` (even if this means an extra conection is created), since we only measure the perf of the `run()` method.  I suppose it would slightly increase other perf metrics like memory and number of TCP connections, but it seems minimal.  If you really want to avoid this, then yes you can use the sync client even in `setupAsync()`.  The framework doesn't care if `setupAsync()` actually blocks instead of using async.

In .NET we only use the async APIs for setup/cleanup, though I don't think this creates an extra TCP connection since .NET sync is just sync-over-async.

If we do think this is really important for Java, we should probably modify the base PerfStressTest to have both `setup()/setupAsync()` and `cleanup()/cleanupAsync()`, to make it more clear that tests should be consistent across setup/run/cleanup.

File: sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/core/ServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        if (CoreUtils.isNullOrEmpty(eventHubName)) {
+            throw new IllegalStateException("Environment variable EVENTHUB_NAME must be set");
+        }
+
+        // Setup the service client
+        EventHubClientBuilder builder = new EventHubClientBuilder()
+            .connectionString(connectionString, eventHubName);
+
+        eventHubProducerAsyncClient = builder.buildAsyncProducerClient();
+        eventHubProducerClient = builder.buildProducerClient();
```
[the remark this answers is not in the document]
Comment (mikeharder) [reply sdk/eventhubs/azure-messaging-eventhubs-perf/src/main/java/com/azure/messaging/eventhubs/perf/core/ServiceTest.java:45]: If these need to be closed, would it make more sense to create them in `setup()` and close in `cleanup()`, rather than creating in the ctor?  Alternatively we could add a `close()` method to the `PerfStressTest` base class, Python has similar.

File: sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/core/ServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        if (CoreUtils.isNullOrEmpty(connectionString)) {
+            throw new IllegalStateException("Environment variable EVENTHUBS_CONNECTION_STRING must be set");
+        }
+
+        if (CoreUtils.isNullOrEmpty(eventHubName)) {
+            System.out.println("Environment variable EVENTHUB_NAME must be set");
+            System.exit(1);
+        }
+
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(4);
```
Comment (mikeharder) [inline sdk/eventhubs/microsoft-azure-eventhubs-perf/src/main/java/com/microsoft/azure/eventhubs/perf/core/ServiceTest.java:35]: Why is this hardcoded to 4 threads?  Should there be an option to control this?

**Review by mikeharder — APPROVED** [review]

**Comment by JonathanGiles** [comment]
@g2vinay Can this be finished off and merged?
